opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp

```diff
@@ -613,8 +619,205 @@ class SimulatorFullyImplicitBlackoilEbos
             }
         }
     }
-
-
+    
+    //Fetch variables required for the well report
+    std::vector< std::vector<double>> getWellReport(const UnitSystem& units,
+                                   const Opm::PhaseUsage& pu,
+                                   const Wells* wells,
+                                   WellState& xw)
+    {
+        const int np = wells->number_of_phases;
+        int nw = wells->number_of_wells;
+        std::vector< std::vector<double>> wr ( 8 , std::vector<double> ( nw, 0.0 ) );
+        enum { Water = BlackoilPhases::Aqua, Oil = BlackoilPhases::Liquid, Gas = BlackoilPhases::Vapour };
+        
+        for (int w = 0; w < nw; ++w) {
+            double wellRate = 0.0;
+            const bool is_producer = wells->type[w] == PRODUCER;       
+            if( pu.phase_used[Water]) {
+                //Water phase rate per well
+                wr[0][w] = abs(xw.wellRates()[np*w + pu.phase_pos[Water]]);
+            }              
+            if( pu.phase_used[Oil]) {
+                //Oil phase rate per well
+                wr[1][w] = abs(xw.wellRates()[np*w + pu.phase_pos[Oil]]);
+            }         
+            if( pu.phase_used[Gas]) {
+                //Gas phase rate per well
+                wr[2][w] = abs(xw.wellRates()[np*w + pu.phase_pos[Gas]]);
+            }
+            //Unit conversion
+            if (units.getType() == UnitSystem::UnitType::UNIT_TYPE_FIELD) { 
+                wr[0][w] = unit::convert::to(wr[0][w],unit::stb/unit::day);
```

Comment:
I think you should separate the "which unit family am I using" functionality from the "which unit type does this quantity have" functionality.

One way to do that is by making a small struct containing the units/conversion factors, for example it could have fields for "gas rate", "liquid rate", "gas-oil ratio" etc. In a separate function you would test for unit system (field, metric etc.) and set the struct, then use it here.

Of course, I think we may already have such functionality in opm-parser! I do not recall exactly where, @joakim-hove could you help?